### ISSUE TITLE
Use async http for reconciliation

### ISSUE THREAD

**Issue by gitonthescene** [OpenRefine/OpenRefine#3657]
Reconciliation can take a long time and that’s not helped by the fact that batches are [run synchronously](https://github.com/OpenRefine/OpenRefine/blob/82b1cf0bfaf3c8ab57a12e0f2004bdeb71432383/main/src/com/google/refine/operations/recon/ReconOperation.java#L282).

### Proposed solution

If we used some [async http library](https://github.com/AsyncHttpClient/async-http-client) we could issue all the requests in parallel and [collect the results](https://www.baeldung.com/java-completablefuture) and leave it to the reconciliation service to handle the requests.  The service should already be configured to handle the traffic whether by queuing up requests or load balancing across servers.  We could also allow the service to optionally configure some throttle in their [service manifest](https://reconciliation-api.github.io/specs/latest/#service-manifest) to allow them to request that clients play nice and/or avoid refusing requests.  We might also want to let the service control the batchSize instead of [hardcoding it](https://github.com/OpenRefine/OpenRefine/blob/14f43dc2cccab09eacf5d9771110cf08e579d82b/main/src/com/google/refine/model/recon/StandardReconConfig.java#L235).

With these changes we could scale up a reconciliation service with load balancing to improve throughput.

### Alternatives considered

There is a limit to what can be done to improve throughput when the client sends requests synchronously especially when the batch size is fixed.  You could make the batchSize much larger and leave it to the service to break up the request and farm it out.  This might be worth doing first as a stop gap.

### Additional context

I think this should be a nice self-contained project which will require a bit more than editing a line or two but probably not a lot more.  Setting up tests could be interesting too depending how you deal with servers refusing high volumes.

**Comment by wetneb** [comment]
I agree the batch size should be configurable: #2549.

I don't think it would be a good default behaviour to "issue all requests in parallel" in general: this is not nice to the reconciliation server even for mid-sized datasets. Say you send batches of 100 queries, on a dataset with 10,000 rows this means 100 HTTP requests: you definitely don't want to fire those in parallel (especially because they will be pretty long-running). Many services already process the reconciliation queries inside a given batch in parallel anyway.

In the new architecture, I have added some limited concurrency: up to N requests are done in parallel, where N is the number of partitions of the project (which would eventually become configurable, but only at project creation time), so it's not quite what you are asking for but again I wouldn't be comfortable with letting people fire off hundreds of parallel requests from one client (especially as maintainer of the Wikidata recon service…)

**Comment by gitonthescene** [comment]
I can understand that.  That's why I suggested it be part of the service manifest to specify how much parallelism they're able to absorb.  If you're doing 100 queries with 10,000 rows that could take a very long time.  If it's your own service, you might want to simply scale up your back end to speed things up.

**Comment by wetneb** [comment]
Yes services could announce which concurrency they recommend, so adding the "reconciliation API design" tag for that.

**Comment by wetneb** [comment]
@gitonthescene if you feel like improving the specs on this, it would be fantastic! https://github.com/reconciliation-api/specs/issues/66

**Comment by wetneb** [comment]
In the 4.0 branch, reconciliation runs 4 queries in parallel by default. This could be made more configurable though.

**Comment by gitonthescene** [comment]
Nice!  Definitely a step in the right direction.  Thanks.

**Comment by wetneb** [comment]
I haven't looked into it in detail, but I suspect 429 handling might be part of the base HTTP client that the reconciliation client uses. I plan to have a closer look when improving error handling there (#3194)
